unattended-upgrade

```diff
@@ -1265,86 +1305,97 @@ def get_dpkg_log_content(logfile_dpkg, install_start_time):
-                      kernel_pkgs_regexp)
-
-        return set([pkg for pkg in auto_removable
-                    if (re.match(kernel_pkgs_regexp, pkg) and
-                        not re.match(running_kernel_version, pkg))])
-    else:
-        logging.debug("APT::VersionedKernelPackages is not set")
-        return set()
+    return {pkg.name for pkg in cache
+            if pkg.is_auto_removable}
 
 
-def is_autoremove_valid(cache, pkgname, auto_removable):
-    # type: (apt.Cache, str, AbstractSet[str]) -> bool
-    changes = set([p.name for p in cache.get_changes()])
+def is_autoremove_valid(cache, pkgname, auto_removable, blacklisted_pkgs,
+                        whitelisted_pkgs):
+    # type: (apt.Cache, str, AbstractSet[str], List[str], List[str]) -> bool
+    changes = cache.get_changes()
     if not changes:
         # package is already removed
         return True
-    if not changes.issubset(auto_removable):
+    pkgnames = {pkg.name for pkg in changes}
+    for pkg in changes:
+        if not is_pkg_change_allowed(pkg, blacklisted_pkgs, whitelisted_pkgs):
+            logging.warning(
+                _("Keeping the following auto-removable package(s) because "
+                  "they include %s which is set to be kept unmodified: %s"),
+                pkg.name, " ".join(sorted(pkgnames)))
+            return False
+    if not pkgnames.issubset(auto_removable):
         if pkgname != "":
             logging.warning(
                 _("Keeping auto-removable %s package(s) because it would"
                   " also remove the following packages which should "
                   "be kept in this step: %s"), pkgname,
-                " ".join(sorted(changes - auto_removable)))
+                " ".join(sorted(pkgnames - auto_removable)))
         else:
             logging.warning(
                 _("Keeping %s auto-removable package(s) because it would"
                   " also remove the following packages which should "
                   "be kept in this step: %s"), len(auto_removable),
-                " ".join(sorted(changes - auto_removable)))
+                " ".join(sorted(pkgnames - auto_removable)))
 
         cache.clear()
         return False
+    for packagename in pkgnames:
+        if cache.running_kernel_pkgs_regexp and \
+           cache.running_kernel_pkgs_regexp.match(packagename):
+            logging.warning(
+                _("Keeping the following auto-removable package(s) because "
+                  "they include %s which package is related to the running "
+                  "kernel: %s"), packagename, " ".join(sorted(pkgnames)))
+            cache.clear()
+            return False
     return True
 
 
-def do_auto_remove(cache, auto_removable, logfile_dpkg, minimal_steps,
-                   verbose=False, dry_run=False):
-    # type: (apt.Cache, AbstractSet[str], str, bool, bool, bool) -> bool
+def do_auto_remove(cache,             # type: apt.Cache
```

Comment:
Nice! Curious does the dep484 stuff helps you? I find it quite nice but I wonder if that is just me :)

Reply:
Sometimes it catches errors which is good, but I prefer when a language automatic type inference with strong types and we don't have to add redundant comments to the code. All in all it still seems to worth the effort here.